### PR TITLE
Add Align to Reference option in timeline

### PR DESCRIPTION
**Pull request by bmatherly**
This is phase 1 of the "Align to Reference" feature.

This phase allows multiple timeline clips to be selected. The user selects the reference track to align to. Clips can only be aligned to a track and not to a specific clip. If a clip is on the reference track, it will be flagged as an error. An undo command allows the entire change to be reverted with undo.

This adds a project link dependency to fftw3 (which is already available in our builds)

Demo:

https://user-images.githubusercontent.com/821968/162655869-dbb9c31f-e4cd-4834-be7c-762867f03c4c.mp4

Future phases:
2) Quality threshold to detect clips that can not be aligned. In Phase 1, unrelated clips will still find some alignment (even though there really is none)
3) Drift compensation - automatically apply a speed factor to a clip to correct cameras that record at different rates (some code for this exists, but it is not ready for testing)

### CODE REVIEW

**Comment by ddennedy** [comment]
Hey, just to let you know I have been waiting to see if I will make a 22.04 bug fixes release before merging this (and my glaxnimate MLT producer).

**Comment by bmatherly** [comment]
> Hey, just to let you know I have been waiting to see if I will make a 22.04 bug fixes release before merging this

Thanks for letting me know. It is OK with me if this request waits for a while. I am currently working on "phase 2". This request does not need to wait for phase 2, but if I finish phase 2 before this PR is merges, then I will add to it.

Do you have any comments about the UI or the usability? I would like to resolve usability comments before it is merged.

**Comment by bmatherly** [comment]
> 2) Quality threshold to detect clips that can not be aligned. 

This is now implemented

File: src/commands/timelinecommands.h [hunk 1]
```diff
@@ -571,6 +571,28 @@ class ReplaceCommand : public QUndoCommand
     UndoHelper m_undoHelper;
 };
 
+
+class AlignCLipsCommand : public QUndoCommand
```
Comment (ddennedy) [inline src/commands/timelinecommands.h:575]: Please change the case of 'L' in Clips to lower case.

File: src/dialogs/alignaudiodialog.cpp [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+{
+    int col = 0;
+    setWindowTitle(title);
+    setWindowModality(QmlApplication::dialogModality());
+
+    QGridLayout* glayout = new QGridLayout();
+    glayout->setHorizontalSpacing(4);
+    glayout->setVerticalSpacing(2);
+    // Track Combo
+    glayout->addWidget(new QLabel(tr("Reference Audio Track")), col, 0, Qt::AlignRight);
```
Comment (ddennedy) [inline src/dialogs/alignaudiodialog.cpp:232]: Only the first word in a field label is supposed to be capitalized: https://shotcut.org/notes/ui-conventions/

File: src/dialogs/alignmentarray.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+    // Ensure the two sequences are transformed
+    transform();
+    from.transform();
+
+    // Calculate the cross-correlation signal
+    for(size_t i = 0; i < m_actualComplexSize; ++i)
+    {
+        correlationBuf[i] = m_forwardBuf[i] * std::conj(from.m_forwardBuf[i]);
+    }
```
Comment (ddennedy) [inline src/dialogs/alignmentarray.cpp:82]: The coding style in this file does not conform. I realize you did not write it, but I want it reformatted.

File: src/dialogs/alignmentarray.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+private:
+    void transform();
+    std::vector<double> m_values;
+    fftw_plan m_forwardPlan;
+    std::complex<double>* m_forwardBuf;
+    double m_autocorrelationMax;
+    size_t m_minimumSize;
+    size_t m_actualComplexSize;
+    bool m_isTransformed = false;
```
Comment (ddennedy) [inline src/dialogs/alignmentarray.h:49]: This is already initialized in the constructor. C++11 allows this in-class initialization, and we allow that in Shotcut. But it is not only mixed but redundant.

File: src/docks/timelinedock.cpp [hunk 5]
```diff
@@ -664,6 +665,15 @@ void TimelineDock::addVideoTrack()
         new Timeline::AddTrackCommand(m_model, true));
 }
 
+void TimelineDock::alignSelectedClips()
+{
+    auto selection = selectionUuids();
+    saveAndClearSelection();
+    AlignAudioDialog dialog(tr("Align clips to audio reference"), &m_model, selection, this);
```
Comment (ddennedy) [inline src/docks/timelinedock.cpp:672]: Window titles should have each word capitalized. Also, if an action triggers a dialog, usually they should have the same string. In `Clip.qml` this is "Align To Reference Track"

**Review by ddennedy — CHANGES_REQUESTED** [review]

File: src/commands/timelinecommands.h (hunk 1 above)
Comment (bmatherly) [reply src/commands/timelinecommands.h]: Done

File: src/dialogs/alignaudiodialog.cpp (hunk 2 above)
Comment (bmatherly) [reply src/dialogs/alignaudiodialog.cpp]: done

File: src/dialogs/alignmentarray.cpp (hunk 3 above)
Comment (bmatherly) [reply src/dialogs/alignmentarray.cpp]: done

File: src/dialogs/alignmentarray.h (hunk 4 above)
Comment (bmatherly) [reply src/dialogs/alignmentarray.h]: done

File: src/docks/timelinedock.cpp (hunk 5 above)
Comment (bmatherly) [reply src/docks/timelinedock.cpp]: done

**Pull request merged**